src/main/java/frc/robot/SwerveModule.java

```diff
@@ -0,0 +1,125 @@
+        public static final double kISteer = 0.0; //TODO: Determine actual kISteer Value
+        public static final double kDSteer = 0.0; //TODO: Determine actual kDSteer Value
+    }
+    private final CANSparkMax mDriveMotor;
+    private final CANSparkMax mSteerMotor;
+    
+    private final RelativeEncoder mDriveEncoder;
+    private final RelativeEncoder mSteerEncoder;
+
+    private final PIDController mSteerPidController;
+
+    private final CANCoder mAbsoluteEncoder;
+    private final boolean mAbsoluteEncoderReversed;
+    private final double mAbsoluteEncoderOffsetRad;
+
+    public SwerveModule(int driveMotorId, int steerMotorId, boolean driveMotorReversed, boolean steerMotorReversed,
+    int absoluteEncoderId, double absoluteEncoderOffset, boolean absoluteEncoderReversed) {
+    //absolute encoders
+    mAbsoluteEncoderOffsetRad = absoluteEncoderOffset;
+    mAbsoluteEncoderReversed = absoluteEncoderReversed;
+    mAbsoluteEncoder = new CANCoder(absoluteEncoderId); //default is degrees per second
+    mAbsoluteEncoder.configFeedbackCoefficient(ModuleConstants.kAbsoluteEncoderCountsPerMin2Rad, "rad", SensorTimeBase.PerSecond); //convert to radians per second
+    //motors
+    mDriveMotor = new CANSparkMax(driveMotorId, MotorType.kBrushless);
+    mSteerMotor = new CANSparkMax(steerMotorId, MotorType.kBrushless);
+    //inverting motors
+    mDriveMotor.setInverted(driveMotorReversed);
+    mSteerMotor.setInverted(steerMotorReversed);
+    //motor encoders
+    mDriveEncoder = mDriveMotor.getEncoder();
+    mSteerEncoder = mSteerMotor.getEncoder();
+
+    //so that we can work with meters and radians instead of rotations
+    mDriveEncoder.getPosition();
+    mDriveEncoder.setPositionConversionFactor(ModuleConstants.kDriveEncoderRotToMeters);
+    mDriveEncoder.setVelocityConversionFactor(ModuleConstants.kDriveEncoderRpm2Mps);
+    mSteerEncoder.setVelocityConversionFactor(ModuleConstants.kSteerEncoderRot2Rad);
+    mSteerEncoder.setVelocityConversionFactor(ModuleConstants.kSteerEncoderRPM2RadPerSec);
+    
+    //TODO: hardware PID?
+    
+    mSteerPidController = new PIDController(ModuleConstants.kPSteer, ModuleConstants.kISteer, ModuleConstants.kDSteer);
+    mSteerPidController.enableContinuousInput(-Math.PI, Math.PI);
+
+    
+    }
+
+    public double getDrivePosition() {
+        return mDriveEncoder.getPosition();
+    }
+
+    public double getSteerPosition() {
+        return mSteerEncoder.getPosition();
+    }
+
+    public double getDriveVelocity() {
+        return mDriveEncoder.getVelocity();
+    }
+    
+    public double getSteerVelocity() {
+        return mSteerEncoder.getVelocity();
+    }
+
+    public double getAbsoluteEncoder() {
+        return mAbsoluteEncoder.getPosition()-mAbsoluteEncoderOffsetRad;
+    }
+
+    public void resetEncoders() {
+        mDriveEncoder.setPosition(0);
+        mSteerEncoder.setPosition(getAbsoluteEncoder());
```

Comment:
You probably should use [this](https://store.ctr-electronics.com/content/api/java/html/classcom_1_1ctre_1_1phoenix_1_1sensors_1_1_c_a_n_coder.html#acd7b0d2b1f0e60ac260772f9351366de), instead of `mSteerEncoder.setPosition(getAbsoluteEncoder());`.